packages/steveo/src/runtime/workflow.ts

```diff
@@ -0,0 +1,378 @@
+import { v4 } from 'uuid';
+import bind from 'lodash.bind';
+import assert from 'node:assert';
+import nullLogger from 'null-logger';
+import take from 'lodash.take';
+import { Step, StepUnknown } from './workflow-step';
+import { IProducer, IRegistry, Logger } from '../common';
+import { WorkflowState } from './workflow-state';
+import { Repositories, Storage } from '../storage/storage';
+import { WorkflowOptions, WorkflowPayload } from '../types/workflow';
+
+interface ExecuteContext {
+  workflowId: string;
+
+  payload: unknown;
+
+  step: StepUnknown;
+
+  state: WorkflowState;
+
+  repos: Repositories;
+}
+
+export class Workflow {
+  $name: string;
+
+  $topic: string;
+
+  $storage: Storage;
+
+  $logger: Logger;
+
+  $registry: IRegistry;
+
+  $producer: IProducer;
+
+  $options: WorkflowOptions;
+
+  /**
+   * The execution step definitions.
+   * Uses an object type as the generic types of the steps
+   * will vary from step to step.
+   */
+  steps: Step<unknown, unknown>[] = [];
+
+  constructor(props: {
+    name: string;
+    topic: string;
+    storage: Storage;
+    logger?: Logger;
+    registry: IRegistry;
+    producer: IProducer;
+    options: WorkflowOptions;
+  }) {
+    this.$name = props.name;
+    this.$topic = props.topic;
+    this.$storage = props.storage;
+    this.$logger = props.logger ?? nullLogger;
+    this.$registry = props.registry;
+    this.$producer = props.producer;
+    this.$options = props.options;
+
+    assert(this.$name, `name must be specified`);
+
+    // Register the workflow, this will make sure we can start a workflow execution
+    this.$registry.addNewTask(this);
+  }
+
+  // Support the existing interface ITask by duck typing members
+  get name() {
+    return this.$name;
+  }
+
+  get topic() {
+    return this.$topic;
+  }
+
+  get options() {
+    return this.$options;
+  }
+
+  /**
+   *
+   */
+  next<State, Result>(step: Step<State, Result>): Workflow {
+    this.steps.push(step as Step<unknown, unknown>);
```

Comment:
this i'd like to see steps as a `Map<string, Step>()` and we would store the step name in the state rather than an integer

Reply:
given the logic below in this file, it might be a substantial change it might be worth a discussion.